src/Compilers/Core/Portable/FileSystem/IFileSystem.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+
+namespace Roslyn.Utilities
+{
+    /// <summary>
+    /// Abstraction over the file system that is useful for test hooks
+    /// </summary>
+    internal interface IFileSystem
```

Comment:
>IFileSystem [](start = 23, length = 11)

I don't quite like this abstraction. Seems like a random set of file system methods. I think it'd be better if each component defined virtual methods that it needs (e.g. RelativePathResolver defines FileExists method).

Reply:
That means that in order to properly mock out a file system I, and every other test out there, have to sub-class many different types in the compiler code base .That's not scalable. It dramatically increases the cost of writing a simple tests. I think it's much better to standardize on a file system abstraction and use that.

Reply:
Fair enough. On the other hand having a single interface will lead to a kitchen sink that adds dependencies where there weren't before. Perhaps we can have IFileSystem that is used for the common compiler set of types, but doesn't bleed thru the rest of the system. `ICommonCompilerFileSystem` would be a better name for it.

Reply:
I'm fine with the name change but I also want to avoid the kitchen sink problem. The intent is for this to be a razor thin wrapper over the existing file system layer. I'm already a bit frustrated about how we calculate the normalized name for files since that requires `FileStream` and that can't be easily mocked in tests. Hoping in the future I can get rid of that to just make this simpler.